crates/libcontainer/src/container/container_kill.rs

```diff
@@ -24,37 +24,87 @@ impl Container {
-        let signal = signal.into().into_raw();
-
-        let pids = if all {
-            let cgroups_path = self.spec()?.cgroup_path;
-            let use_systemd = self
-                .systemd()
-                .context("container state does not contain cgroup manager")?;
-            let cmanger = create_cgroup_manager(&cgroups_path, use_systemd, self.id())?;
-            cmanger.get_all_pids()?
-        } else {
-            vec![self
-                .pid()
-                .context("failed to get the pid of the container")?]
-        };
-
         self.refresh_status()
             .context("failed to refresh container status")?;
         if self.can_kill() {
-            pids.into_iter().try_for_each(|pid| {
-                log::debug!("kill signal {} to {}", signal, pid);
-                signal::kill(pid, signal)
-            })?;
+            self.do_kill(signal, all)?;
+        } else {
+            // just like runc, allow kill --all even if the container is stopped
+            if all && self.status() == ContainerStatus::Stopped {
+                self.do_kill(signal, all)?;
+            } else {
+                bail!(
+                    "{} could not be killed because it was {:?}",
+                    self.id(),
+                    self.status()
+                )
+            }
+        }
+        self.set_status(ContainerStatus::Stopped).save()?;
+        std::process::exit(0);
+    }
 
-            self.set_status(ContainerStatus::Stopped).save()?;
-            std::process::exit(0)
+    pub(crate) fn do_kill<S: Into<Signal>>(&self, signal: S, all: bool) -> Result<()> {
+        if all {
+            self.kill_all_processes(signal)
         } else {
-            bail!(
-                "{} could not be killed because it was {:?}",
+            self.kill_one_process(signal)
+        }
+    }
+
+    fn kill_one_process<S: Into<Signal>>(&self, signal: S) -> Result<()> {
+        let signal = signal.into().into_raw();
+        let pid = self.pid().unwrap();
+
+        log::debug!("kill signal {} to {}", signal, pid);
+        signal::kill(pid, signal)?;
+        // For cgroup V1, a frozon process cannot respond to signals,
+        // so we need to thaw it. Only thaw the cgroup for SIGKILL.
+        if self.status() == ContainerStatus::Paused && signal == signal::Signal::SIGKILL {
+            match get_cgroup_setup()? {
+                libcgroups::common::CgroupSetup::Legacy
+                | libcgroups::common::CgroupSetup::Hybrid => {
+                    let cgroups_path = self.spec()?.cgroup_path;
+                    let use_systemd = self
+                        .systemd()
+                        .context("container state does not contain cgroup manager")?;
+                    let cmanger = create_cgroup_manager(&cgroups_path, use_systemd, self.id())?;
+                    cmanger.freeze(libcgroups::common::FreezerState::Thawed)?;
```

Comment:
Should we leave the cgroup thawed after this? I'm not sure myself. On one hand runc does seem to cgroup thawed for SIGKILL, but then the behavior is inconstant for other signals, which will not thaw the cgroup. If we do choose to re-freeze it, would that make sense, to re-freeze given that the container process is sigkill-ed ; and if so done, how will we handle the race condition between sending the signal, thawing the cgroup and re-freezing it? @utam0k what are your thoughts?

Reply:
To be honest, I'm not sure, but I think there is no problem if we follow runc.
https://github.com/utam0k/runc/blob/2e8b7a1283838d7a8757b5c8a0425b525c28d269/libcontainer/container_linux.go#L381-L383

Reply:
Yeah, I guess that should be fine, as most of the things that use runc (containerd, podman, docker etc) seem to accept this or ignore this anyways.

Reply:
The real world isn't always ideal, maybe you should report the issue to runc.